convert-hf : support bfloat16 conversion

compilade opened the pull request:
As a follow-up to #7075 and #6412, this introduces proper lazy `bfloat16` conversion in `convert-hf-to-gguf.py` with Numpy.

Numpy does not yet support `bfloat16`, but this is still possible.

This implementation, [like the one in `ggml-impl.h`](https://github.com/ggerganov/llama.cpp/blob/4426e2987b566f09c7aa96ada9706cc778637620/ggml-impl.h#L75), makes `nan` quiet, flushes subnormals to zero, and rounds to nearest even value.
This means `bf16` tensor data made with `convert-hf-to-gguf.py` should match exactly what `./quantize` produces from `f32` models.

# Summary of changes

- Unify how lazy tensors work for PyTorch and Numpy
  - I've added `gguf-py/gguf/lazy.py` for this, which defines the `LazyMeta` metaclass and the `LazyBase` base class which is used by both `LazyNumpyTensor` and `LazyTorchTensor`.
    - `LazyTorchTensor` is still defined in `convert-hf-to-gguf.py` to avoid `torch` dependency in `gguf-py`.
  - Lazy Numpy tensors can now support arbitrary expression splits, where one tensor is used in more than one calculation.
  - No more risk of deep recursion, with the use of a `deque` of lazy tensors per expression graph
- `bfloat16` conversion support
  - Add `LlamaFileType` in `gguf-py/gguf/constants.py` to get the correct `ftype` values as in `llama_ftype` from `llama.h`.
    - Not called `GGMLFileType` because it's probably best to reserve this name for an enum analogous to `ggml_ftype`
    - Still open to name suggestions :)
- `--outtype auto` to choose the highest-fidelity 16-bit floating point type according to the type of the first loaded tensor.
  - Uses `f16` if the first tensor has dtype `torch.float16`, and uses `bf16` otherwise, so that `torch.float32` and `torch.bfloat16` tensors keep their range.
- `--outfile` name templating
  - allows using `python3 convert-hf-to-gguf.py --outfile path/to/llama-3-8b-instruct-{ftype}.gguf --outtype auto ./path/to/Meta-Llama-3-8B-Instruct/`, and still get the automatically-chosen output type in the name.

# Testing

> [!NOTE]
>
> The checksum of a model converted with
> ```console
> $ python3 convert-hf-to-gguf.py --outfile ./models/ggml-model-bf16.gguf --outtype bf16 ./path/to/model_dir/
> ```
> and one converted with
> ```console
> $ python3 convert-hf-to-gguf.py --outfile ./models/ggml-model-f32.gguf --outtype f32 ./path/to/model_dir/
> $ ./build/bin/quantize ./models/ggml-model-f32.gguf ./models/ggml-model-bf16.gguf bf16
> ```
> SHOULD EXACTLY MATCH (as of <https://github.com/ggerganov/llama.cpp/pull/7158/commits/95930da30ec55d35565617203af249cecc83ff90>)

- [x] Mamba
  - @compilade Both `bf16` and `f32` quantized to `bf16` (with `./quantize`) result in the same output at `--temp 1` given the same seed with <https://huggingface.co/delphi-suite/v0-mamba-100k>. Model made with `--no-lazy` matches checksum of lazily-converted model. (sha256sum: `272030278cb50b8b1eece85e175e1681c2aeabc430330a38380d2e441099a996`)
- [x] BERT
  - @compilade checksum matches for <https://huggingface.co/BAAI/bge-small-en-v1.5> in `bf16` compared to `./quantize` output from `f32` (sha256sum: `560d1f957d9cb77ddcbbe558cbbb394d18af0500f402be25cb6b2292a3b3cc8a`)
- [x] Llama with MoE
  - @compilade checksum matches for <https://huggingface.co/jtatman/TinyMistral-248m-v2.5-4x-Moe> in `bf16` compared to `./quantize` output from `f32` (sha256sum: `d6e6e1977ffa9cc365d425c107d7a79770b9425cab9db04d695e092fedd00d72`)
- [ ] Other models (TODO, but they should work)

(relevant for at least @jart, @teleprint-me, @bartowski1182)

compilade committed:
convert-hf : support bfloat16 conversion

compilade committed:
gguf-py : flake8 fixes

compilade committed:
convert-hf : add missing space after comma

bartowski1182 commented:
Based on your description, my assumption is that if the original weights are in bf16, you should do convert with outtype bf16 and then everything will just work a bit better. Otherwise no other changes to pipeline (from an end-user perspective) are needed. Is that correct? Or is there even some kind of auto-detection?

compilade committed:
convert-hf : get bit-exact same output as ./quantize

The quantization version was missing.

* convert-hf : don't round bf16 NANs

* convert-hf : save some memory with np.int16 intermediate bf16 weights

* convert-hf : more closely match llama.cpp with which weights to keep in f32

compilade commented:
> Based on your description, my assumption is that if the original weights are in bf16, you should do convert with outtype bf16 and then everything will just work a bit better. Otherwise no other changes to pipeline (from an end-user perspective) are needed. Is that correct? Or is there even some kind of auto-detection?

Yes this is correct. Using `--outtype bf16` should be enough. There is no auto-detection for now.

I just now found a way to make bit-exact identical `bf16` model outputs from `convert-hf-to-gguf.py` compared to `./quantize` from an `f32` model. I'll push it soon (< 1hr). This will make testing for correctness easier (matching checksums should then be sufficient).

Galunid commented on convert-hf-to-gguf.py:
```diff
@@ -2417,8 +2372,8 @@ def parse_args() -> argparse.Namespace:
     parser.add_argument(
-        "--outtype", type=str, choices=["f32", "f16"], default="f16",
-        help="output format - use f32 for float32, f16 for float16",
+        "--outtype", type=str, choices=["f32", "f16", "bf16"], default="f16",
```
Given most models come in bf16 wouldn't it make sense to set it as default?

compilade replied:
The conversion to `bf16` is slightly slower and uses a bit more RAM than `f16` conversion, due to the lack of native Numpy support, so I didn't change the default.

I'll see if I can auto-detect whether the model contains `bf16` tensors (but it will most likely be too complicated). Otherwise, it does make sense to set `bf16` as default if it's very widely used.

ggerganov replied:
We only have rudimentary CPU-only `bf16` support in `ggml`, so `f16` is better for now

colink-snow replied:
when you say CPU-only, I assume you're referring to inference, since all conversion and quantization is currently CPU-only?

colink-snow replied:
> I'll see if I can auto-detect whether the model contains `bf16` tensors (but it will most likely be too complicated). Otherwise, it does make sense to set `bf16` as default if it's widely used.

@compilade could we not attempt to read from config.json? it should have a torch_dtype in it

bartowski1182 replied:
(left comment on wrong account)

>I'll see if I can auto-detect whether the model contains bf16 tensors (but it will most likely be too complicated). Otherwise, it does make sense to set bf16 as default if it's widely used.

@compilade could we not attempt to read from config.json? it should have a torch_dtype in it

Ggerganov, when you say CPU-only, I assume you're referring to inference, since all conversion and quantization is currently CPU-only?

compilade replied:
> could we not attempt to read from config.json? it should have a torch_dtype in it

@bartowski1182 yes, but not all models define that field, so I think a second guess based on the type of the first tensor type in the model will sometimes be necessary. Would you like some kind of `--outtype auto-f16` even if it's not the default?

bartowski1182 replied:
I'd be very interested in that yes, if they set it wrongly I would have probably messed up conversion anyways, but ideally I'd like to be mostly auto bf16 (since so much llama3 is flooding the market) with a tiny bit of intelligence to avoid it if it's explicitly set to something else

To confirm, we're able to quantize from BF16 right? It's not just meant for running at full precision?

compilade replied:
> but ideally I'd like to be mostly auto bf16

Yeah, my idea for `--outtype auto-f16` would use `f16` only if the first tensor of the model has dtype `torch.float16`, and use `bf16` otherwise (so that `bf16` is also chosen when converting from `f32` weights).
Is `auto-bf16` a better name? But it's a ~~bit~~ byte longer... :)
(I think that simply `auto` would be too ambiguous)

> To confirm, we're able to quantize from BF16 right?

Yes, I can confirm quantizing **from** `bf16` works.

compilade replied:
I've added `--outtype auto-f16` to guess which 16-bit float type to use depending on the type of the first loaded tensor (which seems like a good enough heuristic). `f16` is only used if that first tensor has type `torch.float16`, and otherwise, `bf16` is used.

Since most models are converted to then be quantized, I feel like this would be a good default, even if `bfloat16` inference is CPU-only, but I've kept `f16` as default for now.

htdung167 replied:
@compilade I'm using llama.cpp as a black box so I don't understand how it works. Can you explain why bf16 is only used for CPU? Will there be GPU support in the future?

bartowski1182 replied:
@compilade I think '--outtype auto' might be fine, since at its core that's what it's doing

only thing is, and maybe i can do it myself if it's too much tech debt, is it possible to figure out the auto-chosen version earlier and use it for naming the outfile? I also kind of want to add an --out-dir option, since I like putting the conversion in a new location and naming it -f32.gguf/-f16.gguf but won't be able to make it programmatic with this setup

compilade replied:
> Can you explain why bf16 is only used for CPU? Will there be GPU support in the future?

@htdung167 This PR is only about conversion, and the convert script always has been CPU-only.
Inference (text generation) with `bf16` was worked on by @jart in #6412. A relevant comment from there regarding future GPU support would be <https://github.com/ggerganov/llama.cpp/pull/6412#issuecomment-2028967868>.

> I think '--outtype auto' might be fine, since at its core that's what it's doing

@bartowski1182 I agree, I've thought about this more, and I'll change this to `auto` instead of `auto-f16`. I don't think there will be `auto-anything-else` anyway.

>  is it possible to figure out the auto-chosen version earlier and use it for naming the outfile?

yes, this is already possible by checking the logs. It would also be possible to do automatically, but not everyone has the same naming conventions, so maybe the right way to do this would be with a `.format()` pattern? For example, `--outfile llama-3-8b-instruct.{ftype}.gguf`, or `--outfile llama-3-8b-instruct.{outtype}.gguf` or `--outfile llama-3-8b-instruct.{}.gguf`. Not sure which to support (all?), but it should be clearer than `%s`. It would also be possible to allow using `{FTYPE}` or `{OUTTYPE}` for upper-cased type names.

I see you've used `fp16` and `fp32` in the past, but this will use `f16` and `f32`, respectively, for these type names.

bartowski1182 replied:
yeah i've made the transition to f16/f32, was an oversight from me naming them 'fp'

a format option would be amazing

compilade replied:
Should `auto` simply try to be as lossless as possible? Like, if the model is originally in `f32`, make the output `f32`? Or should it always select a 16-bit type?

bartowski1182 replied:
My vote would be on compressing even if originally it was f32, if the person covering wants f32 they'll specify, otherwise presumably they're always converting with the intention of quantizing where it won't matter

jart replied:
"Do no harm" should be the default.

compilade committed:
convert-hf : add --outtype auto-f16

A reason for this to exist is for model quantizers who want an initial
GGUF with the most fidelity to the original model while still using
a 16-bit float type instead of 32-bit floats.

teleprint-me commented on convert-hf-to-gguf.py:
```diff
@@ -215,6 +237,23 @@ def extra_f16_tensors(self, name: str, new_name: str, bid: int | None, n_dims: i
+        # same as ggml_compute_fp32_to_bf16 in ggml-impl.h
+        def np_fp32_to_bf16(n: np.ndarray):
+            # force nan to quiet
+            n = np.where((n & 0x7fffffff) > 0x7f800000, (n & 0xffff0000) | (64 << 16), n)
```
Everything looks good. This is the only line that bugs me. Not sure about the `(64 << 16)` for logical OR. I'm knee deep in a dataset, so my mental state is not 100% there right now. Could be nothing.

compilade replied:
It's doing the equivalent of these lines in `ggml-impl.h`:

https://github.com/ggerganov/llama.cpp/blob/befddd0f15de6efb15d7e7f5b527dfb671f4196f/ggml-impl.h#L82-L85

`(64 << 16)` came from adapting this while postponing the right shift until after rounding.
`(n & 0xffff0000)` is to avoid rounding up the NANs by setting the low bits to zero.

compilade committed:
convert-hf : remove a semicolon because flake8 doesn't like it

It's a reflex from when programming in C/C++, I guess.

compilade committed:
convert-hf : support outtype templating in outfile name

* convert-hf : rename --outtype auto-f16 to --outtype auto

bartowski1182 commented:
This is looking really good. What's the next steps to get this merged? I can do some testing if that's what is needed

compilade commented:
> What's the next steps to get this merged?

Honestly, it's pretty much ready. But the newly-added `--outfile` templates may need to be reviewed.
From my own manual tests the `bf16` conversion works quite well and outputs the exact same files (with matching cheksums) as if converted from `./quantize` from a `f32` model.
If you can find a counter-example I'd be very curious about that.

`--outtype auto` seems to behave well too, automatically choosing either `f16` or `bf16`. The only thing that could still be changed is the default outtype, but I've left it to `f16` for now because it's supported by more backends in `ggml` (see <https://github.com/ggerganov/llama.cpp/pull/7158#discussion_r1595677448>).

Conversion performance (speed-wise) for `bf16` isn't *that* good (around `40-60 MB/s` on my machine), but this is inherent to the lack of native support in Numpy. It can be improved later, since at least this is usable and produces the correct bits. I have some ideas of how to make this faster, even before trying to multi-thread it, like using something else than `np.vectorize`. I have some performance improvements ready (making `bf16` conversion go at `104 MB/s`), but they are tied to other changes that also add `Q8_0` support to `convert-hf-to-gguf.py` (giving identical results as with the reference implementation in `ggml-quants.c`, unlike the current `Q8_0` conversion in `convert.py`) which will be more appropriate in another PR.

If there is no objection, I would like to merge this at 2024-05-11 15:00 UTC.

bartowski1182 commented:
I assume any slowdown in converting to bf16 is made up for by the speed of quanting bf16 instead of f32

Actually on that subject, since we can't inference bf16 with GPU, can we make imatrix for bf16 with GPU?

Looking forward to it! may pull it to try out in the interim

jart commented:
> Actually on that subject, since we can't inference bf16 with GPU

Yet.

bartowski1182 commented:
@jart love the hint haha. Yeah I figured it's coming, but in the meantime I'm curious how it works, is GPU inference support required for imatrix on GPU?
